buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.gradle.plugin;
+
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.elasticsearch.gradle.Version;
+import org.elasticsearch.gradle.VersionProperties;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.InvalidUserDataException;
+import org.gradle.api.Project;
+import org.gradle.api.file.CopySpec;
+import org.gradle.api.plugins.JavaPluginConvention;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.TaskAction;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Creates a plugin descriptor.
+ */
+public class PluginPropertiesTask extends DefaultTask {
+
+    private PluginPropertiesExtension extension;
+
+    private File descriptorOutput = new File(getProject().getBuildDir(),
+            "generated-resources/plugin-descriptor.properties");
+
+    public PluginPropertiesTask() {
+        Project project = this.getProject();
+        this.extension = project.getExtensions().create("esplugin", PluginPropertiesExtension.class, project);
+    }
+
+    @TaskAction
+    public void configurePluginPropertiesTask() {
+        Project project = this.getProject();
+
+        File templateFile = new File(project.getBuildDir(), "templates");
+        templateFile.getParentFile().mkdirs();
+
+        project.copy((cp) -> {
+            cp.from(getClass().getClassLoader().getResource(this.descriptorOutput.getName())).into(templateFile);
```

Comment:
This copying needs to be using a stream like it did before (using getResourceAsStream and copying in a doLast). The zip filesystem java puts over jars has issues. IIRC I had issues with this failing when build-tools is bundled into a jar (ie for an external plugin).

Reply:
This review comment https://github.com/elastic/elasticsearch/pull/39605#discussion_r263697609 by @atorok pointed me in the direction of using `projct.copy` and I thought perhaps I should make use of `from` and `into` inside the `project.copy`. 

I can replace this with with `getResourceAsStream` but, want to be sure that the upcoming changes are not conflicting with earlier review comments. 

Why do we need the `doLast` part?

Reply:
AFAIK, there is no CopySpec method that takes in input stream. So in order to use getResourseAsStream, we need to actually copy the bytes from the stream, to a file (as we did before). Thus we need some arbitrary code to run, in a doLast.

Reply:
@rjernst How does the `doLast` block get triggered? I have the follow task action 
```
@TaskAction
    public void configurePluginPropertiesTask() {
        Project project = this.getProject();
        
        File templateFile = new File(project.getBuildDir(), "templates/" + descriptorOutput.getName());
        templateFile.getParentFile().mkdirs();

         doLast(unused -> {
                InputStream resourceTemplate = this.getClass().getResourceAsStream("/" + descriptorOutput.getName());
                templateFile.getParentFile().mkdirs();
                try {
                    Files.write(templateFile.toPath(), resourceTemplate.readAllBytes());
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            });
            
        project.copy(copyArg -> {
            copyArg.from(templateFile.getParentFile()).include(descriptorOutput.getName())
                .into(descriptorOutput.getParentFile()).expand(this.generateSubstitutions());
            this.getInputs().properties(this.generateSubstitutions());
        });
    }
``` 
and calling `pluginPropertiesTask.configurePluginPropertiesTask();` ignores the `doLast` 😕

Reply:
A couple things here:
* The name of this method is confusing, since it is not configuration; it is the actual action for the task
* Don't use readAllBytes on the input stream. Instead call `Files.copy(resourceTemplate, templateFIle.toPath())`.
* The doLast shouldn't be on the pluginProperties task. Instead, it was previously on an ancillary task whose sole purpose was to copy the template from the build-tools jar into the plugin project. I see now Alpar asked to not do that anymore. That is fine. In this case, the contents of the doLast can simply be removed. The contents should be run just before the project.copy you have.

Reply:
The `@TaskAction` is in the same realm as `doLast` in that each Gradle task has a list of  actions to execute. `doFirst` and `doLast` are ways to add to that list, but like Ryan points it out are not necessary here, or in custom tasks in general as you are implementing the main task action and can freely move code around.